src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php

```diff
@@ -50,11 +50,39 @@ public function process(ContainerBuilder $container)
 
         $serializerDefinition = $container->getDefinition($this->serializerService);
         $serializerDefinition->replaceArgument(0, $normalizers);
+        $this->addDefaultContextParameter($normalizers, $container);
 
         if (!$encoders = $this->findAndSortTaggedServices($this->encoderTag, $container)) {
             throw new RuntimeException(sprintf('You must tag at least one service as "%s" to use the "%s" service.', $this->encoderTag, $this->serializerService));
         }
 
         $serializerDefinition->replaceArgument(1, $encoders);
+        $this->addDefaultContextParameter($encoders, $container);
+    }
+
+    private function addDefaultContextParameter($services, $container)
+    {
+        foreach ($services as $service) {
+            $definition = $container->getDefinition($service);
+            if (!$definition->isAutowired()) {
+                continue;
+            }
+
+            if (null === $class = $definition->getClass()) {
+                continue;
+            }
+
+            $reflection = new \ReflectionClass($class);
+
+            if (null === $constructor = $reflection->getConstructor()) {
+                continue;
+            }
+
+            foreach ($constructor->getParameters() as $arg) {
+                if ('defaultContext' === $arg->name) {
+                    $definition->setArgument('$defaultContext', '%serializer.default_context%');
+                }
+            }
+        }
```

Comment:
I don't love this level of magic. I'd prefer:

A) Just documenting the parameter and allowing people to wire it up manually
B) Creating a `DefaultContext` service that holds the config and passing that around. For the existing classes that are expecting an array, that's fine - we can pass the array to those classes, but make this service available.

Reply:
I prefer B) because it will work with projects providing extra context options such as API Platform, and doing it manually is tedious.

What we could do:

* Introduce a new `ContextInterface` (that can be extended)
* Allow all parameters that currently accept an array to also accept this new interface, and deprecacate passing an array
* Use it both for the "default context" and the context you can pass to methods (it means that this class MUST be safely serializable)

Reply:
Other better option: create a context builder that returns an array. Then it will be consistent with the builder introduced in HttpClient.

Reply:
Discussion about this in #30888